app/controllers/api/vms_controller.rb

```diff
@@ -169,66 +104,31 @@ def add_event_resource(type, id = nil, data = nil)
     end
 
     def reset_resource(type, id = nil, _data = nil)
-      raise BadRequestError, "Must specify an id for resetting a #{type} resource" unless id
-
-      api_action(type, id) do |klass|
-        vm = resource_search(id, type, klass)
-        api_log_info("Resetting #{vm_ident(vm)}")
-
-        result = validate_vm_for_action(vm, "reset")
-        result = reset_vm(vm) if result[:success]
-        result
-      end
+      enqueue_ems_action(type, id, "Resetting", :method_name => "reset", :supports => true)
     end
     central_admin :reset_resource, :reset
 
     def reboot_guest_resource(type, id = nil, _data = nil)
-      raise BadRequestError, "Must specify an id for rebooting a #{type} resource" unless id
-
-      api_action(type, id) do |klass|
-        vm = resource_search(id, type, klass)
-        api_log_info("Rebooting #{vm_ident(vm)}")
-
-        result = validate_vm_for_action(vm, "reboot_guest")
-        result = reboot_guest_vm(vm) if result[:success]
-        result
-      end
+      enqueue_ems_action(type, id, "Rebooting", :method_name => "reboot_guest", :supports => true)
     end
     central_admin :reboot_guest_resource, :reboot_guest
 
     def rename_resource(type, id, data = {})
-      raise BadRequestError, "Must specify an id for renaming a #{type} resource" unless id
-
       new_name = data.blank? ? "" : data["new_name"].strip
-      raise BadRequestError, "Must specify a new_name" if new_name.blank?
-
-      api_action(type, id) do |klass|
-        vm = resource_search(id, type, klass)
-        api_log_info("Renaming #{vm_ident(vm)} to #{new_name}")
 
-        result = validate_vm_for_action(vm, "rename")
-        result = rename_vm(vm, new_name) if result[:success]
-        result
+      enqueue_ems_action(type, id, "Renaming", :method_name => "rename") do
+        raise BadRequestError, "Must specify a new_name" if new_name.blank?
```

Comment:
There are a few instances of these.

I had debated this.

For a request with bad parameters , this will still look up the vm.
But we still want the block to log the result, add the resource link, and depending if it is a single resource, raise the exception or just use a failed action response hash.;

I can dive in and see if there is a dry way to handle this.